applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp

```diff
@@ -0,0 +1,655 @@
+                if (it_node == element_nodes.end())
+                    all_nodes_locked = true;
+            }
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function unlocks element nodes after finite differencing in parallel.
+     * @details 
+     * @param rElement The element to unlock nodes
+     */
+    void UnlockElementNodes(Element& rElement)
+    {
+        KRATOS_TRY
+        // Loop over element nodes
+        for (auto& node_i : rElement.GetGeometry()) 
+            node_i.UnSetLock();
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function retrieves the basis corresponding to element DOFs
+     * @details 
+     * @param rPhiElemental The element to unlock nodes
+     * @param basisIndex The index of the basis to retrieve
+     * @param rElement The element to retrieve basis
+     * @param rElementDofList The elemental dof list
+     * @param rCurrentProcessInfo The current process info instance
+     */
+    void GetPhiElemental(
+        LocalSystemVectorType& rPhiElemental,
+        const std::size_t basisIndex,
+        const Element& rElement,
+        const TElementDofsVectorType& rElementDofList,        
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY
+
+        // Get elemental and nodal DOFs size
+        const auto element_dofs_size = rElementDofList.size();
+        const auto nodal_dof_size = element_dofs_size/rElement.GetGeometry().size();
+        rPhiElemental.resize(element_dofs_size);
+        
+        // Get PhiElemental
+        for(std::size_t i_node = 0; i_node < rElement.GetGeometry().size(); ++i_node)
+        {
+            const auto& r_node = rElement.GetGeometry()[i_node];
+            const auto* p_phi_nodal = &(r_node.GetValue(ROM_BASIS));
+            for (std::size_t i_dof = 0; i_dof < nodal_dof_size; ++i_dof)
+            {
+                const auto& rp_dof = rElementDofList[i_node*nodal_dof_size+i_dof];
+                rPhiElemental[i_node*nodal_dof_size+i_dof] = (*p_phi_nodal)(rCurrentProcessInfo[MAP_PHI].at(rp_dof->GetVariable().Key()), basisIndex);
+            }
+        }
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function performs forward differencing on the element LHS matrix
+     * @details 
+     * @param rElementMatrixDerivative The element to unlock nodes
+     * @param basisIndex The index of the basis for perturbation
+     * @param rElement The element to compute finite differencing
+     * @param rCurrentProcessInfo The current process info instance
+     */
+    void ForwardDifferencingWithBasis(
+         LocalSystemMatrixType& rElementMatrixDerivative,
+         const std::size_t basisIndex,
+         Element&rElement,
+         const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+
+        // Lock element nodes for OMP parallelism
+        this->LockElementNodes(rElement);
+
+        const auto matrix_size = rElementMatrixDerivative.size1();
```

Comment:
```suggestion
        const std::size_t matrix_size = rElementMatrixDerivative.size1();
```

I don't recommend using `auto` for basic types
In this case it is probably fine but the compiler already screwed me multiple times when I expected double and got int or so :)